core/src/main/java/org/apache/iceberg/BaseIncrementalAppendScan.java

```diff
@@ -77,35 +76,52 @@ public CloseableIterable<FileScanTask> planFiles() {
       toSnapshotIdInclusive = context().toSnapshotId();
     } else if (table().currentSnapshot() != null) {
       toSnapshotIdInclusive = table().currentSnapshot().snapshotId();
+    } else if (context().fromSnapshotId() != null) {
+      throw new IllegalArgumentException(
+          "Invalid config: end snapshot is not set, start snapshot is set, and table has no current snapshot");
     } else {
-      throw new IllegalArgumentException("End snapshot is not set and table has no current snapshot");
+      // It is an empty table (no current snapshot). Both from and to snapshots aren't set either.
+      // In this case, listener notification is also skipped
+      return CloseableIterable.empty();
     }
 
-
+    // fromSnapshotIdExclusive can be null. appendsBetween handles null fromSnapshotIdExclusive properly
+    // by finding the oldest ancestor of end snapshot.
+    Long fromSnapshotIdExclusive = context().fromSnapshotId();
```

Comment:
Can you refactor out all the calls to `context().fromSnapshotId()`? I think it would be more readable to have a variable.

Reply:
addressed in PR #4886